Ex1/week5_4.py

```diff
@@ -0,0 +1,98 @@
+import itertools
+import os
+import re
+from bs4 import BeautifulSoup
+
+# Mixed Tools
+
+# -- Regular version:
+
+
+def interleave(*args):
+    """
+    Take an unlimited list of arguments that can be iterate and returns a list of the elements in order
+    for example:
+    Input: 'abc', [1, 2, 3], ('!', '@', '#')
+    Output: ['a', 1, '!', 'b', 2, '@', 'c', 3, '#']
+    :param args: An iterable
+    :return: A list of elements mixed in order
+    """
+    list_of_tuples = []
+    for elem in zip(*args):
+        list_of_tuples.append(elem)
+    return list(sum(list_of_tuples, ()))
+
+
+# -- Generator version:
+
+
+def interleave_generator(*args):
+    """
+    Take an unlimited list of arguments that can be iterate and returns a list of the elements in order
+    using generator
+    for example:
+    Input: 'abc', [1, 2, 3], ('!', '@', '#')
+    Output: ['a', 1, '!', 'b', 2, '@', 'c', 3, '#']
+      :param args: An iterable
+    :return: A list of elements mixed in order
+    """
+    for elem in zip(*args):
+        yield elem
+
+
+my_list = []
+my_list += [item for element in interleave_generator('abc', [1, 2, 3], ('!', '@', '#'))
+            for item in itertools.chain(element)]
+
+
+# Harry is rational but not bad
+
+
+def generate_num_chapter_with_titles():
+    """
+    A generator function that pass on all unordered html files in directory "potter" and for each
+    html file extract from the title the number og chapter and the correct new file name to be rename
+    this file.
+    For example: for the file 1.html that inside of it has the chapter:
+                Chapter 81: Taboo Tradeoffs, Pt 3,
+                but the correct chapter of 1 is:  A Day of Very Low Probability
+                the name of the file will be rename to:  1 A Day of Very Low Probability.html
+                and 81.html will be rename to: 81 Taboo Tradeoffs, Pt 3.html
+    :return: yield num of chapter and new file name
+    """
+    for filename in os.listdir("resources/potter"):
+        with open("resources/potter/" + filename, "r", encoding="UTF8") as file:
+            soup = BeautifulSoup(file, 'html.parser')
```

Comment:
I dont think the most important aspect of this object, the first thing that we need to know to understand its nature,  is the fact that the module that created it contains the word "soup" in its name. what do you think?